### PR TITLE
[jit] define equality for IValue

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35747 [jit] make node hashing use ivalue equality
* #35019 [jit] kill isSameIdentity
* #35005 [jit] kill `shallowEquals`
* **#34986 [jit] define equality for IValue**

Previously we were reluctant to define equality for IValues, as it's not
totally straightforward. But the vacuum this created basically forced
people to define their own equality comparisons for their own purposes.
We have at least 3 in PyTorch itself, and 2 others outside that I know
of.

These implementations are generally wrong, so we should just bite the
bullet and define equality canonically.

Differential Revision: [D20537901](https://our.internmc.facebook.com/intern/diff/D20537901)

### CODE REVIEW

File: aten/src/ATen/core/Dict_inl.h [hunk 1]
```diff
@@ -3,11 +3,37 @@
...
+  // Otherwise, we first compare by identity for efficiency, then by value (see:
+  // [container equality])
+  return _fastEqualsForContainer(lhs, rhs);
+}
+inline bool operator==(const DictImpl& lhs, const DictImpl& rhs) {
+  return lhs.elementTypes.keyType == rhs.elementTypes.keyType &&
+      lhs.elementTypes.valueType == rhs.elementTypes.valueType &&
+      lhs.dict.size() == rhs.dict.size() &&
+      // see: [container equality]
+      std::equal(
```
Comment (zdevito) [inline aten/src/ATen/core/Dict_inl.h]: two equal dictionaries are not necessarily in the same order, but can still be equals. This code seems wrong.

File: aten/src/ATen/core/List_inl.h [hunk 2]
```diff
@@ -4,6 +4,20 @@
...
+namespace detail {
+inline bool operator==(const ListImpl& lhs, const ListImpl& rhs) {
+  return lhs.elementType == rhs.elementType &&
+      lhs.list.size() == rhs.list.size() &&
+      // see: [container equality]
+      std::equal(
+             lhs.list.cbegin(),
+             lhs.list.cend(),
+             rhs.list.cbegin(),
+             [](const auto& lhs, const auto& rhs) {
```
Comment (zdevito) [inline aten/src/ATen/core/List_inl.h]: lambda seems redundant?

File: aten/src/ATen/core/ivalue.cpp [hunk 3]
```diff
@@ -26,6 +35,18 @@ CAFFE2_API c10::intrusive_ptr<ConstantString> ConstantString::create(
...
 
+bool operator==(const ivalue::Tuple& lhs, const ivalue::Tuple& rhs) {
+  return lhs.elements_.size() == rhs.elements_.size() &&
+      // see [container equality]
+      std::equal(
+             lhs.elements_.cbegin(),
+             lhs.elements_.cend(),
+             rhs.elements_.cbegin(),
+             [](const auto& lhs, const auto& rhs) {
+               return _fastEqualsForContainer(lhs, rhs);
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue.cpp]: redundant lambda?

File: aten/src/ATen/core/ivalue.cpp [hunk 4]
```diff
@@ -132,6 +153,101 @@ bool IValue::overlaps(const IValue& rhs) const {
...
+    case Tag::RRef:
+      return rhs.isRRef() && lhs.isAliasOf(rhs);
+    case Tag::Object:
+      return rhs.isObject() && lhs.isAliasOf(rhs);
+    case Tag::PyObject:
+      return rhs.isPyObject() && lhs.isAliasOf(rhs);
+    case Tag::Capsule:
+      return rhs.isCapsule() && lhs.isAliasOf(rhs);
+    case Tag::Uninitialized:
+      TORCH_INTERNAL_ASSERT(
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue.cpp]: But arguably the `equals` function might be used in internal contexts as well. It's not clear this should throw.

File: aten/src/ATen/core/ivalue.cpp [hunk 5]
```diff
@@ -132,6 +153,101 @@ bool IValue::overlaps(const IValue& rhs) const {
...
+}
+
+
+static bool isUndefinedTensor(const IValue& iv) {
+  return iv.isTensor() && !iv.toTensor().defined();
+}
+bool IValue::is(const IValue& rhs) const {
+  const IValue& lhs = *this;
+  // Special handling for undefined tensors:
+  // 1. Undefined_tensor is None and vice versa.
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue.cpp]: This is really surprising, needs more justification if we are going to commit it.

File: aten/src/ATen/core/ivalue.cpp [hunk 6]
```diff
@@ -132,6 +153,101 @@ bool IValue::overlaps(const IValue& rhs) const {
...
+    // Compare the underlying tensor implementations, as we may have two
+    // distinct `at::Tensor` objects pointing to the same impl.
+    return rhs.isTensor() && lhs.toTensor().is_same(rhs.toTensor());
+  }
+
+  if (lhs.is_intrusive_ptr) {
+    TORCH_INTERNAL_ASSERT(rhs.is_intrusive_ptr);
+    return lhs.payload.as_intrusive_ptr == rhs.payload.as_intrusive_ptr;
+  }
+  // TODO should we do this?
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue.cpp]: This is going to stack overflow?

**Review by zdevito — APPROVED** [review]
Looks mostly good. Inline comments where I think something is suspicious.

File: aten/src/ATen/core/Dict_inl.h (hunk 1 above)
Comment (suo) [reply aten/src/ATen/core/Dict_inl.h]: I thought  our dictionaries were ordered?

File: aten/src/ATen/core/ivalue.cpp (hunk 5 above)
Comment (suo) [reply aten/src/ATen/core/ivalue.cpp]: I am reproducing what already exists in `isSameIdentity()`, but didn't dig much further for why it's there. I'll look into it.

File: aten/src/ATen/core/ivalue.cpp (hunk 6 above)
Comment (suo) [reply aten/src/ATen/core/ivalue.cpp]: the implementation of equality doesn't call into `is`. Nevertheless, maybe we should just return false here…once we get to this point it's not really a valid comparison anymore.

File: aten/src/ATen/core/Dict_inl.h (hunk 1 above)
Comment (suo) [reply aten/src/ATen/core/Dict_inl.h]: okay never mind I know what you mean now

**Pull request closed without merging**